pkg/template/jsonpath.go

```diff
@@ -146,3 +146,37 @@ func relaxedJSONPathExpression(pathExpression string) (string, error) {
 func isTektonExpr(expr string) bool {
 	return tektonVar.MatchString(expr)
 }
+
+// findTektonExpressions searches for and returns a slice of
+// all substrings that are wrapped in $()
+func findTektonExpressions(in string) []string {
+	results := []string{}
+
+	// No expressions to return
+	if !strings.Contains(in, "$(") {
+		return results
+	}
+	// Splits string on $( to find potential Tekton expressions
+	maybeExpressions := strings.Split(in, "$(")
```

Comment:
How do we want to handle expressions like `$($())`? Let's also add a test to verify the behavior.

Reply:
yeah, so this is the interesting one :smile: 

currently, we'd return $() i.e. the inner expression

Reply:
`$` is usually the root element in JSONPath though in the kubectl jsonpath template syntax that we are using does not mention $ at all

https://goessner.net/articles/JsonPath/

https://kubernetes.io/docs/reference/kubectl/jsonpath/

Reply:
```
kubectl get pod -o=jsonpath='{$(.items[0])}'

error: error parsing jsonpath {$(.items[0])}, unrecognized character in action: U+0028 '('
```

Reply:
so, IMO we can document that if you have embed `$()`...we'll just use the innermost one e.g. `"$($($(blahblah)))"` will return `$(blahblah)`

what do you think @wlynch @afrittoli ?

Reply:
What happens if you have `$($($(foo))-bar)`?

It might just be easier to just disallow nested `$()` and make it an error.

Reply:
the innermost rule applies and we get `$(foo)`

Reply:
Returning a error is definitely an option though documented/defined behavior for nested `$(` is also an option. Don't feel too strongly about either

Reply:
Ignoring the `-bar` bit feels unexpected to me. I'm leaning towards making it an error.

Reply:
The bar is only ignored as part of the jsonpath expression...when we do the repalcements -> it still shows up as part of the string i,e, if $(foo) evals to "abc" the value of the binding is `$($(abc)-bar)`

Reply:
Ah! Okay, documenting that as a first step sounds good to me then! We can always revisit this later.